app/lib/features/cross_signing/cross_signing.dart

```diff
@@ -1,874 +1,468 @@
-          _onKeyVerificationRequest(event);
+          _verifEmitter.emit('verification.request', event);
           break;
         case 'm.key.verification.ready':
-          _onKeyVerificationReady(event, false);
+          _verifEmitter.emit('verification.ready', event);
           break;
-        case 'm.key.verification.start':
-          _onKeyVerificationStart(event);
+
+        /* sas_verification_handler: after request accepted */
+        case 'm.key.verification.start': // when start button was clicked by other
+        case 'SasState::Started': // when start button was clicked by me
+          _verifEmitter.emit('sas.started', event);
           break;
-        case 'm.key.verification.cancel':
-          _onKeyVerificationCancel(event);
+        case 'm.key.verification.cancel': // when verifier cancelled before verifiee accepts request
+        case 'SasState::Cancelled':
+          _verifEmitter.emit('sas.cancelled', event);
           break;
-        case 'm.key.verification.accept':
-          _onKeyVerificationAccept(event);
+        // case 'm.key.verification.accept':
+        case 'SasState::Accepted':
+          _verifEmitter.emit('sas.accepted', event);
           break;
-        case 'm.key.verification.key':
+        // case 'm.key.verification.key':
         case 'SasState::KeysExchanged':
-          _onKeyVerificationKey(event);
+          _verifEmitter.emit('sas.keys_exchanged', event);
           break;
-        case 'm.key.verification.mac':
-          _onKeyVerificationMac(event);
+        // case 'm.key.verification.mac':
+        case 'SasState::Confirmed':
+          _verifEmitter.emit('sas.confirmed', event);
           break;
-        case 'm.key.verification.done':
-          _onKeyVerificationDone(event);
+        // case 'm.key.verification.done':
+        case 'SasState::Done':
+          _verifEmitter.emit('sas.done', event);
           break;
       }
     });
+    ref.onDispose(() => _verifPoller?.cancel());
   }
 
-  void _onKeyVerificationRequest(VerificationEvent event) {
-    String? flowId = event.flowId();
-    if (flowId == null || _processMap.containsKey(flowId)) {
-      return;
-    }
-    // this case is bob side
-    _processMap[flowId] = VerificationProcess(
-      verifiyingThisDevice: true, // this device is requested for verification
-      stage: 'm.key.verification.request',
-    );
-    acceptingRequest = false;
-    showModalBottomSheet(
-      context: rootNavKey.currentContext!,
-      builder: (BuildContext context) => Container(
-        decoration: BoxDecoration(
-          borderRadius: BorderRadius.circular(15),
-        ),
-        child: _buildOnRequest(context, event, flowId),
-      ),
-      isDismissible: false,
-    );
+  void installEventFilter(BuildContext context) {
+    _verifEmitter.on('verification.init', (_) {
+      onVerificationInit(context);
+    });
+
+    _verifEmitter.on<VerificationEvent>('request.created', (event) {
```

Comment:
using riverpod stream provider would also make this a bit cleaner as it would split the emission and from the viewer in a cleaner fashion ...

Reply:
Fixed by https://github.com/acterglobal/a3/pull/1257/commits/c9c0569cca442daf8549b67deae5e9980bf7d884